src/ray/gcs/subscription_executor.h

```diff
@@ -0,0 +1,85 @@
+#ifndef RAY_GCS_SUBSCRIPTION_EXECUTOR_H
+#define RAY_GCS_SUBSCRIPTION_EXECUTOR_H
+
+#include <atomic>
+#include <mutex>
+#include "ray/gcs/callback.h"
+#include "ray/gcs/tables.h"
+
+namespace ray {
+
+namespace gcs {
+
+/// \class SubscriptionExecutor
+/// SubscriptionExecutor class encapsulates the implementation details of
+/// subscribe/unsubscribe to elements (e.g.: actors or tasks or objects or nodes).
+/// Support subscribe to a specific element or subscribe to all elements.
+template <typename ID, typename Data, typename Table>
+class SubscriptionExecutor {
+ public:
+  SubscriptionExecutor(Table &table) : table_(table) {}
+
+  ~SubscriptionExecutor() {}
+
+  /// Subscribe to operations of all elements.
+  /// Repeated subscription will return a failure.
+  ///
+  /// \param client_id The type of update to listen to. If this is nil, then a
+  /// message for each update will be received. Else, only
+  /// messages for the given client will be received.
+  /// \param subscribe Callback that will be called each time when an element
+  /// is registered or updated.
+  /// \param done Callback that will be called when subscription is complete.
+  /// \return Status
+  Status AsyncSubscribe(const ClientID &client_id,
```

Comment:
Maybe rename it to `AsyncSubscribeAll` to distinguish it with the other method?

Reply:
I think it's ok both way.